nano/node/wallet.cpp

```diff
@@ -1512,6 +1534,67 @@ void nano::wallets::reload ()
 	}
 }
 
+void nano::wallets::do_work_regeneration ()
+{
+	std::unique_lock<std::mutex> regeneration_lock (difficulty_mutex);
+	while (!stopped)
+	{
+		if (!difficulty_reque.empty ())
+		{
+			auto const first (difficulty_reque.begin ());
+			auto now (std::chrono::steady_clock::now ());
+			auto queued (first->first);
+			auto block (first->second);
+			std::shared_ptr<nano::block> block_l (block);
+			bool confirmed (false);
+			difficulty_reque.erase (first);
+			if ((now - queued) >= node.config.work_recalc_inverval)
+			{
+				std::unique_lock<std::mutex> lock (node.active.mutex);
+				auto existing (node.active.roots.find (block->qualified_root ()));
+				if (node.active.roots.end () != existing)
+				{
+					confirmed = existing->election->confirmed.load ();
+				}
+				lock.unlock ();
+				uint64_t difficulty (0);
+				nano::work_validate (*block, &difficulty);
+				if (!confirmed && node.active.active_difficulty < difficulty)
+				{
+					block_l = update_work_action (block);
+				}
+				if (!confirmed)
+				{
+					if (block != block_l)
+					{
+						node.active.update_difficulty (*block_l.get ());
+						node.network.flood_block (block_l);
+					}
+				}
+			}
+			if (!confirmed)
+			{
+				regeneration_lock.unlock ();
+				queue_work_regeneration (now, block_l);
```

Comment:
I might be tempted to move `difficulty_reque.insert (std::make_pair (time_point, block_a));` to another private function which is not thread safe and call that here instead of lock/unlocking (and also call it in `queue_work_regeneration ()`), or even just call `difficulty_reque.insert (std::make_pair (time_point, block_a));` in place (however then you have duplication). These solutions are probably not any cleaner than what you are doing already, but thought I'd suggest some alternatives anyway. Feel free to ignore.